orchagent/routeorch.cpp

```diff
@@ -1206,6 +1226,15 @@ bool RouteOrch::addRoutePost(const RouteBulkContext& ctx, const NextHopGroupKey
     const sai_object_id_t& vrf_id = ctx.vrf_id;
     const IpPrefix& ipPrefix = ctx.ip_prefix;
 
+    if (m_fgNhgOrch->fgNhgPrefixes.find(ipPrefix) != m_fgNhgOrch->fgNhgPrefixes.end()
+            && vrf_id == gVirtualRouterId)
+    {
+        /* Only support the default vrf for Fine Grained ECMP */
+        SWSS_LOG_INFO("Reroute %s:%s to fgNhgOrch", ipPrefix.to_string().c_str(), 
+                nextHops.to_string().c_str());
+        return m_fgNhgOrch->addRoute(vrf_id, ipPrefix, nextHops);
```

Comment:
>m_fgNhgOrch->addRoute [](start = 15, length = 21)

RouteOrch::addRoute() and RouteOrch::addRoutePost will be called once for each route request in RouteOrch::doTask().
So m_fgNhgOrch->addRoute() will be called twice for each route request. Is it intened?

Reply:
Based on my code walk through, 2 calls for the m_fgNhgOrch->addRoute will happen only if the initial fgNhgOrch addRoute failed. So it will function like a retry in failure cases. 

For success case it should return True, and ```consumer.m_toSync.erase(it)``` should occur, as a result it would get skipped when it gets to the main caller of addRoutePost(hence being processed only 1 time):
```
        // Go through the bulker results
        auto it_prev = consumer.m_toSync.begin();
        while (it_prev != it)
```

Do you see this occurring differently/is my understanding off?

Reply:
My understanding is that if the first `m_fgNhgOrch->addRoute` in `RouteOrch::addRoute` succeeded, it will be called again in `RouteOrch::addRoutePost`.

---
In reply to: [487338902](https://github.com/Azure/sonic-swss/pull/1315#discussion_r487338902) [](ancestors = 487338902)

Reply:
That appears to be true if the route is processed by routeorch(because ```RouteOrch::addRoute``` always returns ```FALSE```, hence guaranteeing that it gets processed by addRoutePost too). But if processed by fgNhgOrch it returns ```TRUE``` for success and is not sent to addRoutePost in that case. 

I set loglevel to info and verified that m_fgNhgOrch->addRoute is only being called 1 time in success case.

Reply:
Thanks @anish-n, very clear now.

You mentioned that
> 2 calls for the m_fgNhgOrch->addRoute will happen only if the initial fgNhgOrch addRoute failed. So it will function like a retry in failure cases

This is not the intention of `RouteOrch::addRoutePost`, because it is intended to handle bulk operation results.

So suggest you remove this block in `RouteOrch::addRoutePost`, and the failed route entry in `m_toSync` will be retried in next `doTask()`

---
In reply to: [493819285](https://github.com/Azure/sonic-swss/pull/1315#discussion_r493819285) [](ancestors = 493819285,487338902)

Reply:
Makes sense, sure, will remove the block.